### PR TITLE
Refactoring of Hosted Slots, allows recursive Inner Arcs.

### LINKED ISSUE #2534: Recursive inner Arcs blocked on SlotComposer
Now that inner Arcs are represented as Arc instances (#2524) we could foresee having recursive inner arcs. This is however blocked on SlotComposer. I may not get to it that anytime soon, this refactoring doesn't seem trivial and we don't have a burning need for recursive inner arcs. We will however need that if we use inner arcs for holding real arcs (e.g. social posts), which themselves may do multiplexing. (@shans if he would like to comment on prioritization).

#### Technical Details:

Today creation of a slot in the inner arc is represented as a `HostedSlotConsumer` in the `SlotContext` of the transformation particle. Slots provided by the particle in the inner arc which consumes a hosted slot are represented as `HostedSlotContext`.

This causes problems, as when an a particle in the inner arc (represented in the `SlotComposer` as a `HostedSlotConsumer`) attempts to create a slot in its own inner arc, there is no context against in which to create a new `HostedSlotConsumer` in the recursive innerArc.

The model we have for inner arcs in the `SlotComposer` is very counter intuitive to me, it seems to me that what we should be doing is:

- When a transformation particle creates a slot in the inner arc, this should be represented as a `SlotContext` of some flavor, as it *is* a slot. Today it is represented as `SlotConsumer` - which is weird as no particle is yet connected to that slot.
- Slots provided by the particle in the inner arc should be simply `SlotContexts`, as they are at this point not intercepted by the transformation particle and obey regular rules of slot composing.

@mmandlis if she would like to comment on `SlotComposer` architecture.

### PR DESCRIPTION
**Pull request by piotrswigon**
**Why**: To allow recursive inner arcs and simplify slot composition at the boundaries of arcs.

Fixes #2534 - can be read for why inner arcs were blocked on this refactoring.

**No more HostedSlotConsumer**

HostedSlotConsumer was an entity representing a hosted particle in slot composing. It was an odd duck, as it got created in response to creating a hosted slot. That is - even before the particle that it represented was instantiated. Structurally it was linked to the SlotContext consumed by the transformation particle - which was weird, as the hosted particle was not really consuming that slot, it was consuming the slot created in the inner arc, but the latter was not represented at all in slot composing.

This refactoring removes HostedSlotConsumer - all SlotConsumers are equal now, hosted particles are simply consuming a hosted slot instead of a provided slot.

**HostedSlotContext is the slot created by the transformation particle**

We didn't have a representation of slot created by the transformation in the inner arc. HostedSlotContext was a representation of the slots provided by particles consuming slots created by transformation particles.

HostedSlotContext is a different beast now - it is the representation of the slot created by the transformation particle in the inner arc.

**Implications**

Transformation particle does not need to declare what particle will consume a slot it creates. The slot created in the inner arc can be consumed by multiple hosted particles, just as a regular slot can.

### CODE REVIEW

File: src/runtime/slot-composer.ts [hunk 1]
```diff
@@ -182,10 +161,13 @@ export class SlotComposer {
       // which has been extracted from DOM.
       // TODO: FIXIT!
       if (eventlet.data && eventlet.data.key) {
-        const hostedConsumers = this.consumers.filter(c => c instanceof HostedSlotConsumer && c.transformationSlotConsumer === slotConsumer);
-        for (const hostedConsumer of hostedConsumers) {
-          if (hostedConsumer instanceof HostedSlotConsumer && hostedConsumer.storeId) {
-            const store = hostedConsumer.arc.findStoreById(hostedConsumer.storeId);
+        // We fire off multiple async operations and don't wait.
+        slotConsumer.hostedSlotContexts.forEach(async ctx => {
```
Comment (mmandlis) [inline src/runtime/slot-composer.ts]: nit - i think we agreed to use the conventional for loop instead of forEach

File: src/runtime/slot-dom-consumer.ts [hunk 2]
```diff
@@ -26,13 +27,9 @@ export class SlotDomConsumer extends SlotConsumer {
     this._observer = this._initMutationObserver();
   }
 
-  constructRenderRequest(hostedSlotConsumer): string[] {
+  constructRenderRequest(): string[] {
```
Comment (mmandlis) [inline src/runtime/slot-dom-consumer.ts]: I'm trying to remember why we originally wanted to prefix the hosted-slot templates with the transformation particle prefix, but i think we can no longer do that, is it not needed anymore?
sorry, this is a very vague question, i'll dig it myself later. also there is an issue i file https://github.com/PolymerLabs/arcs/issues/2572 to reconsider the whole way we fetch and cache templates. let's discuss separately.

File: src/runtime/strategies/map-slots.ts [hunk 3]
```diff
@@ -114,6 +113,7 @@ export class MapSlots extends Strategy {
 
   static specMatch(slotConnection, slot) {
     return slotConnection.slotSpec && // if there's no slotSpec, this is just a slot constraint on a verb
+          slot.spec && // if there is no spec on the slot, it is a hosted slot in the inner arc
```
Comment (mmandlis) [inline src/runtime/strategies/map-slots.ts]: maybe a helper method on the slot, like isInnerSlot? the thing here is that (i think) sometimes `slot` is a recipe-slot and sometimes it is a slot-context... maybe worth making separate methods? anyway, this is outside the scope of this PR, just grumbling)

File: src/runtime/slot-context.ts [hunk 4]
```diff
@@ -12,28 +12,108 @@ import {assert} from '../platform/assert-web.js';
...
+ * Represents a slot created by a transformation particle in the inner arc.
+ * 
+ * Render calls for that slot are routed to the transformation particle,
+ * which receives them as innerArcRender calls.
+ * 
+ * TODO:
+ * Today startRender/stopRender calls for particles rendering into this slot are governed by the
+ * availability of the container on the transformation particle. This should be optional and only
+ * used if the purpose of the innerArc is rendering to the outer arc. It should be possible for
+ * the particle which does no rendering at all to create an inner arc with hosted slots, which
```
[the remark this answers is not in the document]
Comment (piotrswigon) [reply src/runtime/slot-context.ts]: I mean doesn't consume a slot, updated for clarity :) Thanks!

File: src/runtime/slot-composer.ts (hunk 1 above)
Comment (piotrswigon) [reply src/runtime/slot-composer.ts]: True. I had to restructure a bit to maintain the fire and forget behavior that I think makes sense here, but done.

File: src/runtime/slot-dom-consumer.ts (hunk 2 above)
Comment (piotrswigon) [reply src/runtime/slot-dom-consumer.ts]: It looks like we never passed anything in the constructRenderRequest(), so this is at least this cleanup is not making anything worse...

File: src/runtime/strategies/map-slots.ts (hunk 3 above)
Comment (piotrswigon) [reply src/runtime/strategies/map-slots.ts]: Alice is refactoring in her upcoming patch so that these 2 are handled separately. Let's see what she proposes :)

**Comment by piotrswigon** [comment]
I've made some edit proposals in https://docs.google.com/document/d/18bqfCPU2fT3ZxQt3Og5FUfm5j6YlhTdgt24x3Sx5AYY/edit#heading=h.2wouwzyxde0l, but I don't have editing rights, so I cannot do diagram updates.

**Comment by mmandlis** [comment]
> I've made some edit proposals in https://docs.google.com/document/d/18bqfCPU2fT3ZxQt3Og5FUfm5j6YlhTdgt24x3Sx5AYY/edit#heading=h.2wouwzyxde0l, but I don't have editing rights, so I cannot do diagram updates.

Sorry for the inconvenience, just gave you edit access and accepted all your suggestions.
